pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java

```diff
@@ -193,18 +196,20 @@ public String getTaskType() {
           continue;
         }
 
-        // Get the bucket size and buffer
+        // Get the bucket size, buffer and number of parallel buckets (1 as default)
         long bucketMs =
-            TimeUtils.convertPeriodToMillis(mergeConfigs.get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY));
+            TimeUtils.convertPeriodToMillis(mergeConfigs.get(MergeTask.BUCKET_TIME_PERIOD_KEY));
         long bufferMs =
-            TimeUtils.convertPeriodToMillis(mergeConfigs.get(MinionConstants.MergeTask.BUFFER_TIME_PERIOD_KEY));
+            TimeUtils.convertPeriodToMillis(mergeConfigs.get(MergeTask.BUFFER_TIME_PERIOD_KEY));
+        int numParallelBuckets = mergeConfigs.get(MergeRollupTask.NUM_PARALLEL_BUCKETS) != null
+            ? Integer.parseInt(mergeConfigs.get(MergeRollupTask.NUM_PARALLEL_BUCKETS)) : DEFAULT_NUM_PARALLEL_BUCKETS;
 
         // Get watermark from MergeRollupTaskMetadata ZNode
-        // windowStartMs = watermarkMs, windowEndMs = windowStartMs + bucketTimeMs
+        // windowStartMs = watermarkMs, windowEndMs = windowStartMs + bucketTimeMs * numParallelBuckets
         long waterMarkMs =
             getWatermarkMs(preSelectedSegments.get(0).getStartTimeMs(), bucketMs, mergeLevel, mergeRollupTaskMetadata);
         long windowStartMs = waterMarkMs;
-        long windowEndMs = windowStartMs + bucketMs;
+        long windowEndMs = windowStartMs + bucketMs * numParallelBuckets;
```

Comment:
+1. I feel the algorithm should be loop over bucket one by one, and keep tracking if there are spill over segments.

Alternatively, we might also be able to put multiple buckets into the same task as long as it follows the `maxNumRecordsPerTask`.

Reply:
Good point, updated to use a helper function to get the `endWindow` which honors buffer time and watermarks, so the tasks are scheduled with `numParallelBuckets` at best effort.

Reply:
> +1. I feel the algorithm should be loop over bucket one by one, and keep tracking if there are spill over segments.
> 
> Alternatively, we might also be able to put multiple buckets into the same task as long as it follows the `maxNumRecordsPerTask`.

Reply:
> +1. I feel the algorithm should be loop over bucket one by one, and keep tracking if there are spill over segments.
> 
> Alternatively, we might also be able to put multiple buckets into the same task as long as it follows the `maxNumRecordsPerTask`.

I'm trying to keep the buckets in the same structure `SegmentSelectionResult` so it can fit in current code easily, the structure will maintain the stats related to spilled over or merged segments.
IMO, it would be simpler and cleaner if one task only handle one time bucket instead of multiple.